### ISSUE TITLE
Static File Server Endpoint

### ISSUE THREAD

**Issue by AndruePeters** [CrowCpp/Crow#497]
Hey,

I've read documentation and looked at some issues, but I'm not sure how to fix this.

``` html
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="css/bootstrap.css" rel="stylesheet" >  
</head>
<body></body>
</html
```

I don't want to have to prefix all links with "static".
I thought if I changed the endpoint prefix, then this would work.

`#define CROW_STATIC_ENDPOINT "/<path>"`

Shouldn't this redirect the endpoint `/path` to `/static/path`? 

This still presents me with 404 errors, though. Is there a way to get the desired behavior? I'm testing multiple servers and really don't want to have to do `static/css/bootstrap.css`.

**Comment by AndruePeters** [comment]
I've been thinking about how to implement this, but it seems quite difficult.

Is there anyway that a special rule can be made, so all files in `static` can use relative paths?

**Comment by AndruePeters** [comment]
This is a current workaround I have, but it's not great:

``` cpp
    CROW_ROUTE(app, "/css/<string>")
    ([&app](const std::string& path) {
        crow::response res;
        res.body = getFileContents(getStaticPath("css", path));
        res.add_header("Content-Type", "stylesheet");
        return res;
    });

    CROW_ROUTE(app, "/js/<string>")
    ([](const std::string& path) {
        crow::response res;
        res.body = getFileContents(getStaticPath("js", path));
        res.add_header("Content-Type", "text/javascript");
        return res;
    });
```

**Comment by The-EDev** [comment]
Changing `CROW_STATIC_ENDPOINT` only changes the endpoint that needs to be requested, if you want to change where your files are on the server, you need to change `CROW_STATIC_DIRECTORY`.
<br><br>
> Is there anyway that a special rule can be made, so all files in static can use relative paths?

This is mostly up to the browser to do that, so a URL like `https://example.com/index.html` with a reference to `css/stylesheet.css`. Your browser would request `https://example.com/css/stylesheet.css`, which you can get by setting `CROW_STATIC_DIRECTORY` to `""`.

For subdirectories like `https://example.com/category/index.html` you can set up a blueprint with a custom static dir (though I might have noticed a bug while writing this comment).
<br><br>
> This is a current workaround I have, but it's not great

I can see a few issues with this implementation. For starters, if you're going to use a response it would be best to use `([](crow::response& res) {` and `res.end()` rather than defining it later and then using `return res`.

Another thing is using `<string>` instead of `<path>` disqualifies any `/` characters, making any files inside subdirectories invalid.

Lastly is that you can use Crow's `response::set_static_file_info(path)` which handles the `Content-Type`, sanitization (to prevent unauthorized navigation), and returning a 404 if the file doesn't exist.

**Comment by AndruePeters** [comment]
Thanks for you response. I'll take a look at more of this tomorrow, but using the route with `response::set_static_file_info(path)` didn't work. The page hung for a while. This was essentially what I had, but the page refused to finish loading.

I didn't see blueprints when I went through the website docs. I'll take a look at that. Also, thanks for `CROW_STATIC_DIRECTORY`. I'll take a look at that.

``` cpp
    CROW_ROUTE(app, "/css/<string>")
    ([&app](const std::string& path) {
        crow::response res;
        res.response::set_static_file_info("/static/" + path;)
        return res;
    });
    ```

**Comment by The-EDev** [comment]
> The page hung for a while. This was essentially what I had, but the page refused to finish loading.

I'm guessing this is due to #469. It's been fixed in `master` but v1.0 still has it. I hope this helps.

**Comment by The-EDev** [comment]
I'll close this for the time being. @AndruePeters let me know if you're facing issues still.
